aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -107,36 +107,59 @@ Tensor pinverse(const Tensor& self, double rcond) {
   return at::matmul(V.conj() * S_pseudoinv.unsqueeze(-2), U.transpose(-2, -1).conj());
 }
 
-static inline Tensor _matrix_rank_helper(const Tensor& self, bool symmetric) {
+Tensor linalg_matrix_rank(const Tensor& self, optional<double> tol, bool hermitian) {
+  // Matrices or batch of matrices are allowed
+  TORCH_CHECK(self.dim() >= 2, "linalg_matrix_rank: Expected as input a matrix or a batch of matrices, but got a tensor of size: ", self.sizes());
+
+  // NumPy doesn't take into account possible input with no elements and it errors on max not defined for this case
+  // Let's output 0 for this case, since that kind of matrices have zero number of non-zero rows, hence rank is 0.
+  if (self.numel() == 0) {
+    auto result_shape = self.sizes().vec();
+    result_shape.pop_back();
+    result_shape.pop_back();  // result's shape is equal to self.shape[0:-2]
+    return at::zeros(result_shape, self.options().dtype(ScalarType::Long));
+  }
+
+  // We compute matrix rank as the number of singular or absolute eigen values above 'tol' threshold
   Tensor S;
-  if (!symmetric) {
+  if (!hermitian) {
     Tensor U, V;
+    // TODO: replace self.svd with linalg_svd
     std::tie(U, S, V) = self.svd(/*some=*/true, /*compute_uv=*/false);
   } else {
     Tensor eigvecs;
+    // TODO: replace self.symeig with linalg_eigh
     std::tie(S, eigvecs) = self.symeig(/*eigenvectors=*/false);
     S = S.abs();
   }
-  return S;
+
+  if (tol.has_value()) {
+    double tol_value = tol.value();
+    return (S > tol_value).sum(/*dim=*/-1);
+  } else {
+    ScalarType real_dtype = toValueType(typeMetaToScalarType(self.dtype()));
+    double tol_value = _get_epsilon(real_dtype) * std::max(self.size(0), self.size(1));
+    Tensor max_S = std::get<0>(S.max(/*dim=*/-1));
+    return (S > max_S.mul_(tol_value).unsqueeze_(-1)).sum(/*dim=*/-1);
+  }
 }
 
-Tensor matrix_rank(const Tensor& self, double tol, bool symmetric) {
-  TORCH_CHECK((at::isFloatingType(self.scalar_type()) || at::isComplexType(self.scalar_type())) && self.dim() == 2,
-              "matrix_rank(", self.scalar_type(), "{", self.sizes(), "}): expected a 2D tensor "
-              "of floating types");
+Tensor& linalg_matrix_rank_out(Tensor& result, const Tensor& self, optional<double> tol, bool hermitian) {
```

Comment:
Add a todo here that we should update this to not copy into result to save space